### ISSUE TITLE
Usage with mkdocs-snippet-plugin

### ISSUE THREAD

**Issue by stiller-leser** [fralau/mkdocs_macros_plugin#68]
Hi, 

I am trying to use this plugin together with https://github.com/mprivat/mkdocs-snippet-plugin .

When including in `mkdocs.yml` like this:
```yaml
plugins:
  - macros:
      verbose: true
  - snippet:
      base_path: doc
```
the macros-plugin works, but throws 
``` 
INFO    -  [macros] - ERROR # _Macro Rendering Error_

**UndefinedError**: 'snippet' is undefined

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/mkdocs_macros/plugin.py", line 443, in render
    return md_template.render(**page_variables)
  File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/usr/local/lib/python3.9/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/usr/local/lib/python3.9/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 1, in top-level template code
jinja2.exceptions.UndefinedError: 'snippet' is undefined
``` 

When including it like this:
```yaml
plugins:
  - snippet:
      base_path: doc
  - macros:
      verbose: true
```
then the macros plugin does not work at all.

Is there a way to delegate certain function calls to other plugins?

Best regards,
stiller-leser

**Comment by fralau** [comment]
The short answer is that that both mkdocs-macros and [mkdocs-git-snippet](https://github.com/mercari/mkdocs-git-snippet) use a jinja2 template engine ([here is its call to jinja2](https://github.com/mercari/mkdocs-git-snippet/blob/903ccfcbac21b7e31a36f09b17a511bb18914df8/mkdocs_git_snippet/plugin.py#L5)). You must delete one of the two from your `mkdocs.yml` file.

**In more detail**: With your first `mkdocs.yml` file, hat mkdocs-macros will not find the macro `gitsnippet()` and will fail. In the second, mkdocs-git-snippet will not find your macros and will fail.

The idea is that mkdocs-macros is generic macro tool, whereas mkdocs-git-snippet is a specialized tool. So the common sense, would be... to implement the functionality of mkdocs-git-snippet in mkdocs-macros.

I'm not exactly sure of what [the code of `gitsnippet()` ](https://github.com/mercari/mkdocs-git-snippet/blob/903ccfcbac21b7e31a36f09b17a511bb18914df8/mkdocs_git_snippet/plugin.py#L5)is doing (the doc doesn't explain too well and the code does not contain any comments).

1. But you would have to rewrite it within a macro module (`main.py`). If you understand the gitsnippet's code, that should be straightforward.
2. Or better, the plugin's author (@mercari) or yourself could reimplement this function as a [pluglet](https://mkdocs-macros-plugin.readthedocs.io/en/latest/pluglets/), so that anybody could easily install it on top of mkdocs-macros.

Basically you would extract the working pieces from the plugin, and fit them again in mkdocs-macros's framework. **Writing a macro or pluglet is substantially easier and faster than writing a full plugin** (that's the whole purpose 😄  ).

Another approach, is that you **explain what problem you are trying to solve with `gitsnippet()`**? So that (perhaps) it can be solved directly with pieces that mkdocs-macros offers out-of-the-box. Mkdocs-macros offers the `git` object, which should be accessible in a python module (as `env.variables.git`).

**Comment by stiller-leser** [comment]
Hi @fralau,

first off: Thank you for that detailed description. This is really appreciated.

To be honest, we are currently evaluating your plugin, so this is not yet a blocker for us. We are using the snippet plugin to pull in parts of markdown files from various GitHub repositories into document files in a central mkdocs-repo (where some additional content is hosted). As I understood the plugin it is checking out the specified git repos and is then copying the respective markdown snippets to their new location (in our case in the central repository). So I suppose you could describe it as a "remote-include directive".

It seems that this indeed would be fairly easy to implement here or simply could be ported as a complete pluglet (as suggested by you).

**Comment by fralau** [comment]
Thanks a lot for your answer. Every description of a use case is useful.
Yes, I believe it would be fairly easy to implement that with mkdocs-macros.

In general, most plugins that rely on jinja2 would be incompatible with each other.

🤦 Oh, and I forgot: there is something you could do, at leat with mkdocs-macros: **change the markers**, i.e. replace the the mustaches ``{{}}``by something else (e.g. ``[[ ]]``) ! [See the description here](https://mkdocs-macros-plugin.readthedocs.io/en/latest/advanced/#how-to-prevent-interpretation-of-jinja-like-statements), under Solution 3.

This was requested at some point, by a user who had such a case. 

Of course, if you go this way, it might be difficult to turn back.

**Comment by fralau** [comment]
Or else, you could fork or tweak mkdocs-git-snippet, so that it uses different markers, on demand. Either way.

**Comment by fralau** [comment]
@stiller-leser How does this issue seem to you now?

**Comment by stiller-leser** [comment]
We convinced our client to change his workflow. So for now we decided not to use the macros plug in. But thank you for the quick and detailed response to this issue.

Have a great weekend!

Holen Sie sich Outlook für Android<https://aka.ms/AAb9ysg>

________________________________
